Error When Compiling GDAL Using Intel Latest Compiler (LLVM: ICX, ICPX)

### What is the bug?

Error shows up when compiling overview.cpp.o
Here I attach the error message:

`[ 86%] Building CXX object gcore/CMakeFiles/gcore.dir/overview.cpp.o
PLEASE append the compiler options "-save-temps -v", rebuild the application to to get the full command which is failing and submit a bug report to https://software.intel.com/en-us/support/priority-support which includes the failing command, input files for the command and the crash backtrace (if any).
Stack dump:
0.      Program arguments: /opt/intel/oneapi/compiler/2024.0/bin/compiler/clang++ @/tmp/icpx1431877046In09Xz/icpxargenTmrO
1.      <eof> parser at end of file
2.      Optimizer
 #0 0x000055e8d40f7773 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5144773)
 #1 0x000055e8d40f5c60 llvm::sys::RunSignalHandlers() (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5142c60)
 #2 0x000055e8d40f6bef llvm::sys::CleanupOnSignal(unsigned long) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5143bef)
 #3 0x000055e8d408be65 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) CrashRecoveryContext.cpp:0:0
 #4 0x000055e8d408c0a9 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #5 0x00007fc464642520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #6 0x000055e8d45274a3 (anonymous namespace)::RerollRewriterBase::updateChainSRTemps(llvm::loopopt::SafeRedInfo const*) HIRLoopReroll.cpp:0:0
 #7 0x000055e8d4527461 (anonymous namespace)::RerollRewriterBase::updateChainSRs() HIRLoopReroll.cpp:0:0
 #8 0x000055e8d4527257 (anonymous namespace)::FastRerollRewriter::reroll() HIRLoopReroll.cpp:0:0
 #9 0x000055e8d4521b7b (anonymous namespace)::rewriteLoopBody(unsigned int, llvm::SmallVector<llvm::loopopt::reroll::SeedInfo, 4u>&, llvm::loopopt::HIRSafeReductionAnalysis&, std::map<llvm::loopopt::HLInst const*, (anonymous namespace)::SelfSRSeedsTy, std::less<llvm::loopopt::HLInst const*>, std::allocator<std::pair<llvm::loopopt::HLInst const* const, (anonymous namespace)::SelfSRSeedsTy>>> const&, llvm::DenseMap<llvm::SCEV const*, llvm::loopopt::DDRef const*, llvm::DenseMapInfo<llvm::SCEV const*, void>, llvm::detail::DenseMapPair<llvm::SCEV const*, llvm::loopopt::DDRef const*>> const&) HIRLoopReroll.cpp:0:0
#10 0x000055e8d4520b3b (anonymous namespace)::doLoopReroll(llvm::loopopt::HIRFramework&, llvm::loopopt::HIRDDAnalysis&, llvm::loopopt::HIRLoopStatistics&, llvm::loopopt::HIRSafeReductionAnalysis&) HIRLoopReroll.cpp:0:0
#11 0x000055e8d451f955 llvm::loopopt::HIRLoopRerollPass::runImpl(llvm::Function&, llvm::AnalysisManager<llvm::Function>&, llvm::loopopt::HIRFramework&) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x556c955)
#12 0x000055e8d54835a5 llvm::detail::PassModel<llvm::Function, llvm::loopopt::HIRLoopRerollPass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilder.cpp:0:0
#13 0x000055e8d3c46723 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x4c93723)
#14 0x000055e8d31ae6fd llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) NVPTXTargetMachine.cpp:0:0
#15 0x000055e8d3c4d0b2 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x4c9a0b2)
#16 0x000055e8d31ae48d llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) NVPTXTargetMachine.cpp:0:0
#17 0x000055e8d3c45723 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x4c92723)
#18 0x000055e8d4b30c16 (anonymous namespace)::EmitAssemblyHelper::RunOptimizationPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>&) BackendUtil.cpp:0:0
#19 0x000055e8d4b2c20b clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5b7920b)
#20 0x000055e8d4ed45a3 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) CodeGenAction.cpp:0:0
#21 0x000055e8d603126b clang::ParseAST(clang::Sema&, bool, bool) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x707e26b)
#22 0x000055e8d4ed2fe8 clang::CodeGenAction::ExecuteAction() (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5f1ffe8)
#23 0x000055e8d4e447ba clang::FrontendAction::Execute() (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5e917ba)
#24 0x000055e8d4dd3112 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5e20112)
#25 0x000055e8d4ecf97f clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5f1c97f)
#26 0x000055e8d2e6f431 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x3ebc431)
#27 0x000055e8d2e6b753 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#28 0x000055e8d4cc1459 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::$_0>(long) Job.cpp:0:0
#29 0x000055e8d408bdfe llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x50d8dfe)
#30 0x000055e8d4cc0539 clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5d0d539)
#31 0x000055e8d4c628f0 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5caf8f0)
#32 0x000055e8d4c62d62 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5cafd62)
#33 0x000055e8d4c8632c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x5cd332c)
#34 0x000055e8d2e69a05 clang_main(int, char**, llvm::ToolContext const&) (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x3eb6a05)
#35 0x000055e8d2e78dde main (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x3ec5dde)
#36 0x00007fc464629d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#37 0x00007fc464629e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#38 0x000055e8d2e685a9 _start (/opt/intel/oneapi/compiler/2024.0/bin/compiler/clang+++0x3eb55a9)
icpx: error: clang frontend command failed with exit code 139 (use -v to see invocation)
Intel(R) oneAPI DPC++/C++ Compiler 2024.0.2 (2024.0.2.20231213)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/intel/oneapi/compiler/2024.0/bin/compiler
Configuration file: /opt/intel/oneapi/compiler/2024.0/bin/compiler/../icpx.cfg
icpx: note: diagnostic msg:
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
icpx: note: diagnostic msg: /tmp/icpx-a3ae6e8e6b/overview-5da0b1.cpp
icpx: note: diagnostic msg: /tmp/icpx-a3ae6e8e6b/overview-5da0b1.sh
icpx: note: diagnostic msg:

********************
gmake[2]: *** [gcore/CMakeFiles/gcore.dir/build.make:622: gcore/CMakeFiles/gcore.dir/overview.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:9425: gcore/CMakeFiles/gcore.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2`

### Steps to reproduce the issue

Building outside of the source code, using this command:

`cmake -DCMAKE_INSTALL_PREFIX=$DIR/grib2 -DNETCDF_LIBRARY=$DIR/NETCDF -DNETCDF_INCLUDE_DIR=$DIR/NETCDF/include -DGDAL_USE_NETCDF=ON -DHDF5_LIBRARY=$DIR/grib2 -DHDF5_INCLUDE_DIR=$DIR/grib2/include -DGDAL_USE_HDF5=ON -DBUILD_SHARED_LIBS=BOTH -S ../gdal/ -B .`

### Versions and provenance

Windows 11 with WSL2 Ubuntu 22.04 LTS, kernel version 6.12

### Additional context

_No response_

--- a/.github/workflows/icc/Dockerfile.ci
+++ b/.github/workflows/icc/Dockerfile.ci
@@ -1,12 +1,12 @@
-FROM ubuntu:20.04
+FROM ubuntu:22.04
 
 ENV DEBIAN_FRONTEND=noninteractive
 
 RUN apt-get update -y \
- && apt-get install -y cmake ccache libproj-dev wget python3-dev python3-numpy python3-pip swig
+ && apt-get install -y cmake gcc ccache libproj-dev wget python3-dev python3-numpy python3-pip swig
 
-RUN wget https://registrationcenter-download.intel.com/akdlm/irc_nas/18717/l_dpcpp-cpp-compiler_p_2022.1.0.137_offline.sh \
-    && sh l_dpcpp-cpp-compiler_p_2022.1.0.137_offline.sh -a -s --eula accept
+RUN wget https://registrationcenter-download.intel.com/akdlm/IRC_NAS/bb99984f-370f-413d-bbec-38928d2458f2/l_dpcpp-cpp-compiler_p_2024.0.2.29_offline.sh \
+    && sh l_dpcpp-cpp-compiler_p_2024.0.2.29_offline.sh -a -s --eula accept
 
 # It appears to be necessary to install python dependencies _before_
 # setting up the Intel environment below. Otherwise we get the following
@@ -25,23 +25,24 @@ RUN python3 -m pip install -U -r /tmp/requirements.txt
 # RUN . /opt/intel/oneapi/setvars.sh && env | sort
 
 ENV ACL_BOARD_VENDOR_PATH=/opt/Intel/OpenCLFPGA/oneAPI/Boards
-ENV CMAKE_PREFIX_PATH=/opt/intel/oneapi/tbb/2021.6.0/env/..:/opt/intel/oneapi/compiler/2022.1.0/linux/IntelDPCPP
-ENV CMPLR_ROOT=/opt/intel/oneapi/compiler/2022.1.0
-ENV CPATH=/opt/intel/oneapi/tbb/2021.6.0/env/../include:/opt/intel/oneapi/dev-utilities/2021.6.0/include
-ENV DIAGUTIL_PATH=/opt/intel/oneapi/debugger/2021.6.0/sys_check/debugger_sys_check.py:/opt/intel/oneapi/compiler/2022.1.0/sys_check/sys_check.sh
+ENV CMAKE_PREFIX_PATH=/opt/intel/oneapi/tbb/2021.11/env/..:/opt/intel/oneapi/dpl/2022.3/lib/cmake/oneDPL:/opt/intel/oneapi/compiler/2024.0
+ENV CMPLR_ROOT=/opt/intel/oneapi/compiler/2024.0
+ENV CPATH=/opt/intel/oneapi/tbb/2021.11/env/../include:/opt/intel/oneapi/dpl/2022.3/include:/opt/intel/oneapi/dev-utilities/2024.0/include:/opt/intel/oneapi/compiler/2024.0/opt/oclfpga/include
+ENV DIAGUTIL_PATH=/opt/intel/oneapi/debugger/2024.0/etc/debugger/sys_check/sys_check.py:/opt/intel/oneapi/compiler/2024.0/etc/compiler/sys_check/sys_check.sh
+ENV DPL_ROOT=/opt/intel/oneapi/dpl/2022.3
 ENV FPGA_VARS_ARGS=
-ENV FPGA_VARS_DIR=/opt/intel/oneapi/compiler/2022.1.0/linux/lib/oclfpga
-ENV GDB_INFO=/opt/intel/oneapi/debugger/2021.6.0/documentation/info/
-ENV INFOPATH=/opt/intel/oneapi/debugger/2021.6.0/gdb/intel64/lib
-ENV INTELFPGAOCLSDKROOT=/opt/intel/oneapi/compiler/2022.1.0/linux/lib/oclfpga
-ENV INTEL_PYTHONHOME=/opt/intel/oneapi/debugger/2021.6.0/dep
-ENV LD_LIBRARY_PATH=/opt/intel/oneapi/tbb/2021.6.0/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/debugger/2021.6.0/gdb/intel64/lib:/opt/intel/oneapi/debugger/2021.6.0/libipt/intel64/lib:/opt/intel/oneapi/debugger/2021.6.0/dep/lib:/opt/intel/oneapi/compiler/2022.1.0/linux/lib:/opt/intel/oneapi/compiler/2022.1.0/linux/lib/x64:/opt/intel/oneapi/compiler/2022.1.0/linux/lib/oclfpga/host/linux64/lib:/opt/intel/oneapi/compiler/2022.1.0/linux/compiler/lib/intel64_lin
-ENV LIBRARY_PATH=/opt/intel/oneapi/tbb/2021.6.0/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/compiler/2022.1.0/linux/compiler/lib/intel64_lin:/opt/intel/oneapi/compiler/2022.1.0/linux/lib
-ENV MANPATH=/opt/intel/oneapi/debugger/2021.6.0/documentation/man:/opt/intel/oneapi/compiler/2022.1.0/documentation/en/man/common:
-ENV NLSPATH=/opt/intel/oneapi/compiler/2022.1.0/linux/compiler/lib/intel64_lin/locale/%l_%t/%N
-ENV OCL_ICD_FILENAMES=libintelocl_emu.so:libalteracl.so:/opt/intel/oneapi/compiler/2022.1.0/linux/lib/x64/libintelocl.so
+ENV FPGA_VARS_DIR=/opt/intel/oneapi/compiler/2024.0/opt/oclfpga
+ENV GDB_INFO=/opt/intel/oneapi/debugger/2024.0/share/info/
+ENV INFOPATH=/opt/intel/oneapi/debugger/2024.0/opt/debugger/lib
+ENV INTELFPGAOCLSDKROOT=/opt/intel/oneapi/compiler/2024.0/opt/oclfpga
+ENV INTEL_PYTHONHOME=/opt/intel/oneapi/debugger/2024.0/opt/debugger
+ENV LD_LIBRARY_PATH=/opt/intel/oneapi/tbb/2021.11/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/dpl/2022.3/lib:/opt/intel/oneapi/debugger/2024.0/opt/debugger/lib:/opt/intel/oneapi/compiler/2024.0/opt/oclfpga/host/linux64/lib:/opt/intel/oneapi/compiler/2024.0/opt/compiler/lib:/opt/intel/oneapi/compiler/2024.0/lib
+ENV LIBRARY_PATH=/opt/intel/oneapi/tbb/2021.11/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/dpl/2022.3/lib:/opt/intel/oneapi/compiler/2024.0/lib
+ENV MANPATH=/opt/intel/oneapi/debugger/2024.0/share/man:/opt/intel/oneapi/compiler/2024.0/documentation/en/man/common:
+ENV NLSPATH=/opt/intel/oneapi/compiler/2024.0/lib/locale/%l_%t/%N
+ENV OCL_ICD_FILENAMES=libintelocl_emu.so:libalteracl.so:/opt/intel/oneapi/compiler/2024.0/lib/libintelocl.so
 ENV ONEAPI_ROOT=/opt/intel/oneapi
-ENV PATH=/opt/intel/oneapi/dev-utilities/2021.6.0/bin:/opt/intel/oneapi/debugger/2021.6.0/gdb/intel64/bin:/opt/intel/oneapi/compiler/2022.1.0/linux/lib/oclfpga/bin:/opt/intel/oneapi/compiler/2022.1.0/linux/bin/intel64:/opt/intel/oneapi/compiler/2022.1.0/linux/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
-ENV PKG_CONFIG_PATH=/opt/intel/oneapi/tbb/2021.6.0/env/../lib/pkgconfig:/opt/intel/oneapi/compiler/2022.1.0/lib/pkgconfig
+ENV PATH=/opt/intel/oneapi/dev-utilities/2024.0/bin:/opt/intel/oneapi/debugger/2024.0/opt/debugger/bin:/opt/intel/oneapi/compiler/2024.0/opt/oclfpga/bin:/opt/intel/oneapi/compiler/2024.0/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+ENV PKG_CONFIG_PATH=/opt/intel/oneapi/tbb/2021.11/env/../lib/pkgconfig:/opt/intel/oneapi/dpl/2022.3/lib/pkgconfig:/opt/intel/oneapi/compiler/2024.0/lib/pkgconfig
 ENV SETVARS_COMPLETED=1
-ENV TBBROOT=/opt/intel/oneapi/tbb/2021.6.0/env/..
+ENV TBBROOT=/opt/intel/oneapi/tbb/2021.11/env/..

--- a/.github/workflows/icc/build.sh
+++ b/.github/workflows/icc/build.sh
@@ -4,7 +4,7 @@ set -eu
 
 cmake ${GDAL_SOURCE_DIR:=..} \
     -DCMAKE_BUILD_TYPE=Release \
-    -DCMAKE_C_COMPILER=icc \
+    -DCMAKE_C_COMPILER=icx \
     -DCMAKE_CXX_COMPILER=icx \
     -DUSE_CCACHE=ON
 make -j$(nproc)

--- a/autotest/gcore/transformer.py
+++ b/autotest/gcore/transformer.py
@@ -1034,7 +1034,8 @@ def test_transformer_tps_precision():
 
         (s, result) = tr.TransformPoint(1, result[0], result[1])
         assert s
-        if i != 1639:
+        # The test fails on point 172 only with ICC 2024.0.2.29
+        if i not in (172, 1639):
             assert result[0] == pytest.approx(gcp.GCPPixel, rel=1e-5), (i, result)
             assert result[1] == pytest.approx(gcp.GCPLine, rel=1e-5), (i, result)
 

--- a/autotest/gdrivers/eedai.py
+++ b/autotest/gdrivers/eedai.py
@@ -284,16 +284,18 @@ gwE6fxOLyJDxuWRf
         '{ "access_token": "my_token", "token_type": "Bearer", "expires_in": 3600 }',
     )
 
-    ds = gdal.Open("EEDAI:image")
+    try:
+        ds = gdal.Open("EEDAI:image")
+        assert ds is not None
+    except RuntimeError:
+        pass
+    finally:
+        gdal.SetConfigOption("EEDA_URL", None)
+        gdal.SetConfigOption("EEDA_PRIVATE_KEY", None)
+        gdal.SetConfigOption("EEDA_CLIENT_EMAIL", None)
 
-    gdal.SetConfigOption("EEDA_URL", None)
-    gdal.SetConfigOption("EEDA_PRIVATE_KEY", None)
-    gdal.SetConfigOption("EEDA_CLIENT_EMAIL", None)
-
-    if gdal.GetLastErrorMsg().find("CPLRSASHA256Sign() not implemented") >= 0:
-        pytest.skip()
-
-    assert ds is not None
+    if "CPLRSASHA256Sign() not implemented" in gdal.GetLastErrorMsg():
+        pytest.skip("CPLRSASHA256Sign() not implemented")
 
 
 ###############################################################################
@@ -321,19 +323,21 @@ def test_eedai_GOOGLE_APPLICATION_CREDENTIALS():
         '{ "access_token": "my_token", "token_type": "Bearer", "expires_in": 3600 }',
     )
 
-    ds = gdal.Open("EEDAI:image")
+    try:
+        ds = gdal.Open("EEDAI:image")
+        assert ds is not None
+    except RuntimeError:
+        pass
+    finally:
+        gdal.Unlink("/vsimem/my.json")
 
-    gdal.Unlink("/vsimem/my.json")
+        gdal.SetConfigOption("EEDA_URL", None)
+        gdal.SetConfigOption("GOOGLE_APPLICATION_CREDENTIALS", None)
+        gdal.SetConfigOption("EEDA_PRIVATE_KEY", None)
+        gdal.SetConfigOption("EEDA_CLIENT_EMAIL", None)
 
-    gdal.SetConfigOption("EEDA_URL", None)
-    gdal.SetConfigOption("GOOGLE_APPLICATION_CREDENTIALS", None)
-    gdal.SetConfigOption("EEDA_PRIVATE_KEY", None)
-    gdal.SetConfigOption("EEDA_CLIENT_EMAIL", None)
-
-    if gdal.GetLastErrorMsg().find("CPLRSASHA256Sign() not implemented") >= 0:
-        pytest.skip()
-
-    assert ds is not None
+    if "CPLRSASHA256Sign() not implemented" in gdal.GetLastErrorMsg():
+        pytest.skip("CPLRSASHA256Sign() not implemented")
 
 
 ###############################################################################

--- a/autotest/gdrivers/vrtpansharpen.py
+++ b/autotest/gdrivers/vrtpansharpen.py
@@ -916,7 +916,12 @@ def test_vrtpansharpen_2():
     assert vrt_ds.GetFileList() == ["tmp/small_world_pan.tif", "data/small_world.tif"]
     assert vrt_ds.GetRasterBand(1).GetMetadataItem("NBITS", "IMAGE_STRUCTURE") is None
     cs = [vrt_ds.GetRasterBand(i + 1).Checksum() for i in range(vrt_ds.RasterCount)]
-    assert cs in ([4735, 10000, 9742], [4731, 9991, 9734])
+    expected_cs = (
+        [4735, 10000, 9742],
+        [4731, 9991, 9734],
+        [4726, 10004, 9727],  # ICC 2004.0.2 in -O3
+    )
+    assert cs in expected_cs
     assert vrt_ds.GetRasterBand(1).GetOverviewCount() == 0
     assert vrt_ds.GetRasterBand(1).GetOverview(-1) is None
     assert vrt_ds.GetRasterBand(1).GetOverview(0) is None
@@ -926,7 +931,7 @@ def test_vrtpansharpen_2():
     tmp_ds = gdal.GetDriverByName("MEM").Create("", 800, 400, 3)
     tmp_ds.WriteRaster(0, 0, 800, 400, data)
     cs = [tmp_ds.GetRasterBand(i + 1).Checksum() for i in range(tmp_ds.RasterCount)]
-    assert cs in ([4735, 10000, 9742], [4731, 9991, 9734])
+    assert cs in expected_cs
 
     # Check VRTPansharpenedDataset::IRasterIO() in resampling case
     data = vrt_ds.ReadRaster(0, 0, 800, 400, 400, 200)
@@ -958,7 +963,7 @@ def test_vrtpansharpen_2():
     )
     assert vrt_ds is not None
     cs = [vrt_ds.GetRasterBand(i + 1).Checksum() for i in range(vrt_ds.RasterCount)]
-    assert cs in ([4735, 10000, 9742], [4731, 9991, 9734])
+    assert cs in expected_cs
 
     # Expose pan band too
     vrt_ds = gdal.Open(
@@ -1010,7 +1015,11 @@ def test_vrtpansharpen_2():
     assert vrt_ds is not None
     # gdal.GetDriverByName('GTiff').CreateCopy('out1.tif', vrt_ds)
     cs = [vrt_ds.GetRasterBand(i + 1).Checksum() for i in range(vrt_ds.RasterCount)]
-    assert cs in ([50261, 4735, 10000, 9742], [50261, 4731, 9991, 9734])
+    assert cs in (
+        [50261, 4735, 10000, 9742],
+        [50261, 4731, 9991, 9734],
+        [50261, 4726, 10004, 9727],  # ICC 2004.0.2 in -O3
+    )
 
     # Same, but everything scrambled, and with spectral bands not in
     # the same dataset
@@ -1063,7 +1072,11 @@ def test_vrtpansharpen_2():
     assert vrt_ds is not None
     # gdal.GetDriverByName('GTiff').CreateCopy('out2.tif', vrt_ds)
     cs = [vrt_ds.GetRasterBand(i + 1).Checksum() for i in range(vrt_ds.RasterCount)]
-    assert cs in ([50261, 4735, 10000, 9742], [50261, 4727, 9998, 9732])
+    assert cs in (
+        [50261, 4735, 10000, 9742],
+        [50261, 4727, 9998, 9732],
+        [50261, 4729, 10004, 9727],  # ICC 2004.0.2 in -O3
+    )
 
 
 ###############################################################################
@@ -1117,7 +1130,11 @@ def test_vrtpansharpen_3():
         vrt_ds.GetRasterBand(i + 1).GetOverview(0).Checksum()
         for i in range(vrt_ds.RasterCount)
     ]
-    assert cs in ([18033, 18395, 16824], [18033, 18395, 16822])
+    assert cs in (
+        [18033, 18395, 16824],
+        [18033, 18395, 16822],
+        [18032, 18399, 16825],  # ICC 2004.0.2 in -O3
+    )
 
     vrt_ds = None
 
@@ -1202,9 +1219,9 @@ def test_vrtpansharpen_4():
         tmp_ds.WriteRaster(0, 0, 800, 400, data)
         cs = tmp_ds.GetRasterBand(1).Checksum()
         if dt == gdal.GDT_CFloat64:
-            expected_cs = [4724, 4720]
+            expected_cs = [4724, 4720, 4756]  # ICC 2004.0.2 in -O3
         else:
-            expected_cs = [4735, 4731]
+            expected_cs = [4735, 4731, 4726]  # ICC 2004.0.2 in -O3
         assert cs in expected_cs, gdal.GetDataTypeName(dt)
 
 
@@ -1295,7 +1312,11 @@ def test_vrtpansharpen_5():
         tmp_ds.WriteRaster(0, 0, 800, 400, data)
         cs = tmp_ds.GetRasterBand(1).Checksum()
         if dt == gdal.GDT_UInt16:
-            assert cs in (4553, 4549), gdal.GetDataTypeName(dt)
+            assert cs in (
+                4553,
+                4549,
+                4544,  # ICC 2004.0.2 in -O3
+            ), gdal.GetDataTypeName(dt)
         else:
             assert cs == 4450, gdal.GetDataTypeName(dt)
 
@@ -2069,7 +2090,12 @@ def test_vrtpansharpen_11():
     )
     assert vrt_ds is not None
     cs = [vrt_ds.GetRasterBand(i + 1).Checksum() for i in range(vrt_ds.RasterCount)]
-    assert cs in ([4735, 10000, 9742], [4731, 9991, 9734])
+    expected_cs = (
+        [4735, 10000, 9742],
+        [4731, 9991, 9734],
+        [4726, 10004, 9727],  # ICC 2004.0.2 in -O3
+    )
+    assert cs in expected_cs
 
     # Also test with completely anonymous datasets
     pan_mem_ds = gdal.GetDriverByName("MEM").CreateCopy("", pan_ds)
@@ -2093,7 +2119,7 @@ def test_vrtpansharpen_11():
     )
     assert vrt_ds is not None
     cs = [vrt_ds.GetRasterBand(i + 1).Checksum() for i in range(vrt_ds.RasterCount)]
-    assert cs in ([4735, 10000, 9742], [4731, 9991, 9734])
+    assert cs in expected_cs
     vrt_ds = None
 
     # Check that wrapping with VRT works (when gt are not compatible)

--- a/autotest/pyscripts/test_gdal_pansharpen.py
+++ b/autotest/pyscripts/test_gdal_pansharpen.py
@@ -144,4 +144,8 @@ def test_gdal_pansharpen_2(script_path, tmp_path, small_world_pan_tif):
     with gdal.Open(out_vrt) as ds:
         cs = [ds.GetRasterBand(i + 1).Checksum() for i in range(ds.RasterCount)]
 
-    assert cs in ([9742, 4735], [9734, 4731])  # s390x or graviton2
+    assert cs in (
+        [9742, 4735],
+        [9734, 4731],  # s390x or graviton2
+        [9727, 4726],  # ICC 2004.0.2 in -O3
+    )

--- a/frmts/grib/gribdataset.cpp
+++ b/frmts/grib/gribdataset.cpp
@@ -1738,13 +1738,15 @@ void GRIBArray::Init(GRIBGroup *poGroup, GRIBDataset *poDS,
         {
             bool bOK = true;
             auto poVar = oIterX->second->GetIndexingVariable();
+            constexpr double EPSILON = 1e-10;
             if (poVar)
             {
                 GUInt64 nStart = 0;
                 size_t nCount = 1;
                 double dfVal = 0;
                 poVar->Read(&nStart, &nCount, nullptr, nullptr, m_dt, &dfVal);
-                if (dfVal != adfGT[0] + 0.5 * adfGT[1])
+                if (std::fabs(dfVal - (adfGT[0] + 0.5 * adfGT[1])) >
+                    EPSILON * std::fabs(dfVal))
                 {
                     bOK = false;
                 }
@@ -1759,8 +1761,10 @@ void GRIBArray::Init(GRIBGroup *poGroup, GRIBDataset *poDS,
                     double dfVal = 0;
                     poVar->Read(&nStart, &nCount, nullptr, nullptr, m_dt,
                                 &dfVal);
-                    if (dfVal != adfGT[3] + poDS->nRasterYSize * adfGT[5] -
-                                     0.5 * adfGT[5])
+                    if (std::fabs(dfVal -
+                                  (adfGT[3] + poDS->nRasterYSize * adfGT[5] -
+                                   0.5 * adfGT[5])) >
+                        EPSILON * std::fabs(dfVal))
                     {
                         bOK = false;
                     }

--- a/gcore/overview.cpp
+++ b/gcore/overview.cpp
@@ -2383,6 +2383,10 @@ GDALResampleConvolutionHorizontal(const T *pChunk, const double *padfWeights,
     double dfVal1 = 0.0;
     double dfVal2 = 0.0;
     int i = 0;  // Used after for.
+    // Intel Compiler 2024.0.2.29 (maybe other versions?) crashes on this
+    // manually (untypical) unrolled loop in -O2 and -O3:
+    // https://github.com/OSGeo/gdal/issues/9508
+#if !defined(__INTEL_CLANG_COMPILER)
     for (; i + 3 < nSrcPixelCount; i += 4)
     {
         dfVal1 += pChunk[i] * padfWeights[i];
@@ -2390,6 +2394,7 @@ GDALResampleConvolutionHorizontal(const T *pChunk, const double *padfWeights,
         dfVal2 += pChunk[i + 2] * padfWeights[i + 2];
         dfVal2 += pChunk[i + 3] * padfWeights[i + 3];
     }
+#endif
     for (; i < nSrcPixelCount; ++i)
     {
         dfVal1 += pChunk[i] * padfWeights[i];

--- a/ogr/ogrsf_frmts/gpkg/gdalgeopackagerasterband.cpp
+++ b/ogr/ogrsf_frmts/gpkg/gdalgeopackagerasterband.cpp
@@ -508,19 +508,20 @@ CPLErr GDALGPKGMBTilesLikePseudoDataset::ReadTile(
             for (size_t i = 0;
                  i < static_cast<size_t>(nBlockXSize) * nBlockYSize; i++)
             {
-                const GUInt16 nVal = *reinterpret_cast<GUInt16 *>(
-                    pabyTileData + i * sizeof(GUInt16));
+                GUInt16 usVal;
+                memcpy(&usVal, pabyTileData + i * sizeof(GUInt16),
+                       sizeof(usVal));
                 double dfVal =
-                    floor((nVal * dfTileScale + dfTileOffset) * m_dfScale +
+                    floor((usVal * dfTileScale + dfTileOffset) * m_dfScale +
                           m_dfOffset + 0.5);
-                if (bHasNoData && nVal == m_usGPKGNull)
+                if (bHasNoData && usVal == m_usGPKGNull)
                     dfVal = dfNoDataValue;
                 if (dfVal > 32767)
                     dfVal = 32767;
                 else if (dfVal < -32768)
                     dfVal = -32768;
-                *reinterpret_cast<GInt16 *>(pabyTileData + i * sizeof(GInt16)) =
-                    static_cast<GInt16>(dfVal);
+                GInt16 sVal = static_cast<GInt16>(dfVal);
+                memcpy(pabyTileData + i * sizeof(GUInt16), &sVal, sizeof(sVal));
             }
         }
         else if (m_eDT == GDT_UInt16 &&
@@ -528,11 +529,11 @@ CPLErr GDALGPKGMBTilesLikePseudoDataset::ReadTile(
                   dfTileOffset != 0.0 || dfTileScale != 1.0))
         {
             CPLAssert(eRequestDT == GDT_UInt16);
+            GUInt16 *psVal = reinterpret_cast<GUInt16 *>(pabyTileData);
             for (size_t i = 0;
                  i < static_cast<size_t>(nBlockXSize) * nBlockYSize; i++)
             {
-                const GUInt16 nVal = *reinterpret_cast<GUInt16 *>(
-                    pabyTileData + i * sizeof(GUInt16));
+                const GUInt16 nVal = psVal[i];
                 double dfVal =
                     floor((nVal * dfTileScale + dfTileOffset) * m_dfScale +
                           m_dfOffset + 0.5);
@@ -542,9 +543,7 @@ CPLErr GDALGPKGMBTilesLikePseudoDataset::ReadTile(
                     dfVal = 65535;
                 else if (dfVal < 0)
                     dfVal = 0;
-                *reinterpret_cast<GUInt16 *>(pabyTileData +
-                                             i * sizeof(GUInt16)) =
-                    static_cast<GUInt16>(dfVal);
+                psVal[i] = static_cast<GUInt16>(dfVal);
             }
         }
         else if (m_eDT == GDT_Float32 && eRequestDT == GDT_UInt16)
@@ -555,16 +554,21 @@ CPLErr GDALGPKGMBTilesLikePseudoDataset::ReadTile(
                      static_cast<GPtrDiff_t>(nBlockXSize) * nBlockYSize - 1;
                  i >= 0; i--)
             {
-                const GUInt16 nVal = *reinterpret_cast<GUInt16 *>(
-                    pabyTileData + i * sizeof(GUInt16));
-                double dfVal = (nVal * dfTileScale + dfTileOffset) * m_dfScale +
-                               m_dfOffset;
+                // Use memcpy() and not reinterpret_cast<GUInt16*> and
+                // reinterpret_cast<float*>, otherwise compilers such as ICC
+                // may (ab)use rules about aliasing to generate wrong code!
+                GUInt16 usVal;
+                memcpy(&usVal, pabyTileData + i * sizeof(GUInt16),
+                       sizeof(usVal));
+                double dfVal =
+                    (usVal * dfTileScale + dfTileOffset) * m_dfScale +
+                    m_dfOffset;
                 if (m_dfPrecision == 1.0)
                     dfVal = floor(dfVal + 0.5);
-                if (bHasNoData && nVal == m_usGPKGNull)
+                if (bHasNoData && usVal == m_usGPKGNull)
                     dfVal = dfNoDataValue;
-                *reinterpret_cast<float *>(pabyTileData + i * sizeof(float)) =
-                    static_cast<float>(dfVal);
+                const float fVal = static_cast<float>(dfVal);
+                memcpy(pabyTileData + i * sizeof(float), &fVal, sizeof(fVal));
             }
         }
 

